File: problems/Swift/299.猜数字游戏.swift
/**
 * 299.猜数字游戏 [Medium]
 * Time Complexity: O(N), Space Complexity: O(C)
 * https://leetcode-cn.com/problems/bulls-and-cows/
 */

class Solution {
  func getHint(_ secret: String, _ guess: String) -> String {
    let secrets = Array(secret)
    let guesses = Array(guess)
    var bulls = 0
    var cows = 0
    var map: [String: Int] = [:]
    for i in 0 ..< secrets.count {
      if secrets[i] == guesses[i] {
        bulls += 1
      } else {
        let key = String(secrets[i])
        if let value = map[key] {
          map[key] = value + 1
        } else {
          map[key] = 1
        }
      }
    }

    for i in 0 ..< guesses.count {
      if secrets[i] != guesses[i] {
        let key = String(guesses[i])
        if let value = map[key], value != 0 {
          map[key] = value - 1
          cows += 1
        }
      }
    }
    return "\(bulls)A\(cows)B"
  }
}

File: problems/rust-solutions/src/299.猜数字游戏.rs
// 299.猜数字游戏 [Medium]
// Time Complexity: O(N), Space Complexity: O(C)
// https://leetcode-cn.com/problems/bulls-and-cows/

impl Solution {
  pub fn get_hint(secret: String, guess: String) -> String {
    let secrets = secret.as_bytes();
    let guesses = guess.as_bytes();
    let mut bulls = 0;
    let mut cows = 0;
    let mut arr: Vec<i32> = vec![0; 10];

    for i in 0..secrets.len() {
      if secrets[i] == guesses[i] {
        bulls += 1;
      } else {
        let secret_num = usize::from(secrets[i] - b'0');
        if (arr[secret_num] < 0) {
          cows += 1;
        }
        arr[secret_num] += 1;

        let guess_num = usize::from(guesses[i] - b'0');
        if (arr[guess_num] > 0) {
          cows += 1;
        }

        arr[guess_num] -= 1;
      }
    }

    return format!("{}A{}B", bulls, cows);
  }
}
